src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Pipeline/Steps/ExceptionEventsPipelineStep.cs

```diff
@@ -40,12 +41,64 @@ public void Invoke(Exception exception)
             // the events once the listener is registered to effectively "catch up".
             if (_eventSource.IsEnabled())
             {
-                ulong identifier = _identifierCache.GetOrAdd(new ExceptionIdentifier(exception));
+                ReadOnlySpan<StackFrame> stackFrames = ComputeEffectiveCallStack(exception);
 
-                _eventSource.ExceptionInstance(identifier, exception.Message);
+                ulong identifier = _identifierCache.GetOrAdd(new ExceptionIdentifier(exception, stackFrames));
+
+                ulong[] frameIds = _identifierCache.GetOrAdd(stackFrames);
+
+                _eventSource.ExceptionInstance(
+                    identifier,
+                    exception.Message,
+                    frameIds);
             }
 
             _next(exception);
         }
+
+        private static ReadOnlySpan<StackFrame> ComputeEffectiveCallStack(Exception exception)
```

Comment:
Let's check what happens for first chance exceptions that will be fatal.

Reply:
We are able to observe unhandled exceptions from the UnhandledException event in the app domain. However, the call stack computation here doesn't work for an exception that was already thrown from within the context of an async method execution. The call stack is preserved on the exception and doesn't match the current thread's call stack because of the async state machine behavior.

This isn't really an issue because we ignore rethrown exceptions.